ImportM3u songs online bugfix-features

TheStef56 opened the pull request:
<!-- Hey there. Thank you so much for improving OuterTune, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

### What is it?

- [x] New feature (user facing)
- [ ] Update to existing feature (user facing)
- [x] Bugfix (user facing)
- [ ] Codebase improvements or refactors (dev facing)
- [ ] Other

### Description of the changes in your PR

In this pr I:

- Fixed: the ImportM3uDialog (it would overflow, not showing buttons). Now it has its dedicated page (screen).
- Fixed: YouTube query (it was still uri-encoded, generating wrong results) .
- Fixed: when you created a new playlist to add the import songs, it would show the uri as default name. Now it shows the proper file.m3u base name as default.
- Improved online search import speed (up to 5-10 times faster, using coroutines).
- Added youtubeSuggestions on top of youtube normal queries to get better matches on online search of m3u imports.
- Added the abilty to change (or remove) one or more of the imported songs before adding to a playlist (in case the first result picked by youtube is not correct).
- Added a percentage that shows the progress.
- Added the ability to import songs as "liked songs".
- Generally improved the UI involved in this process.

NOTE: I added 20 random strings that would result in inexistent songs, just to test rejected songs dropdown.

### Before/After Screenshots/Screen Record

- Before:

https://github.com/user-attachments/assets/98b1a3ff-7d37-46af-9ec7-cae7548b4b19



- After:



https://github.com/user-attachments/assets/c2aa32a8-66f7-4250-a9c3-fa3d96136287




## Due diligence

<!-- Please mark WIP pull requests and "Draft" and only "Ready for review" once it is ready to be merged  -->

- [x] I have read and agreed to the [contribution guidelines](https://github.com/OuterTune/OuterTune/blob/dev/CONTRIBUTING.md).

### Merging strategy / Merge conflict resolution

Select only ONE. If you select none, or both, the first selection will used as your final choice

- [x] When merging this pull request, or in the event of merge conflicts, I give permission for the merger to rebase,
  or squash my code, or apply merge conflict amendments as needed
- [ ] When merging this pull request, or in the event of merge conflicts, I ***DO NOT*** give permission for the
  merger to modify my code to solve merge conflicts. I understand in the event of a merge conflict, I will be
  responsible to resolve merge conflicts in a way that adheres to
  the [contribution guidelines](https://github.com/OuterTune/OuterTune/blob/dev/CONTRIBUTING.md)

<!-- This pull request template is based on Newpipe's:  https://github.com/TeamNewPipe/NewPipe/ -->

mikooomich commented:
Ah I see where you're going with this. I think this should be its own dedicated m3u import page. I initially put this in a dialog because it was rather simple, but if you want to have edit and search features, then it would be a lot nicer for the user (and code wise) to have a page. Then you can have easier dynamic content loading, and don't need those hacks for search, or confine yourself to that dinky little dialog.

Lmk if you want to do it that way, and/or need help with that, ui layout design, etc

TheStef56 commented:
> Ah I see where you're going with this. I think this should be its own dedicated m3u import page. I initially put this in a dialog because it was rather simple, but if you want to have edit and search features, then it would be a lot nicer for the user (and code wise) to have a page. Then you can have easier dynamic content loading, and don't need those hacks for search, or confine yourself to that dinky little dialog.
> 
> Lmk if you want to do it that way, and/or need help with that, ui layout design, etc

I would have implemented it that way, but I wanted to stay as close as the UI originally was. I began making this changes to:

1. Make it work at least (since the last version I had forked wasn't working anymore, youtube changed api I think). 
2. Make it nicer for the user, trying to contribute on the project and suggest my vision of things.

That said, I don't have much time to contribute to this project, and the fork I made satisfies me (and I think most users) enough. If you want, you are free to merge this pr and change it however you want, but in the next future I will be too busy to contribute. This said, I think you should merge to at least let the people Import songs. Thanks for reviewing.

TheStef56 commented:
@mikooomich Ok I'm not liking my implementation either, I will make ImportM3uDialog its own page

TheStef56 commented:
@mikooomich  I switched from a dialog to a proper Screen

TheStef56 commented:
@mikooomich I updated the results in the video preview. Tell me what you think about it and what should I change to make this pr possible to merge.

mikooomich commented:
Its a good start, I'll a bit busy in the next couple weeks, so I'll give it a proper look a bit later. However, I think we should have a different ui design (will send a sketch later today)

TheStef56 commented:
> Its a good start, I'll a bit busy in the next couple weeks, so I'll give it a proper look a bit later. However, I think we should have a different ui design (will send a sketch later today)

Ok ok, thanks for your time.

TheStef56 commented:
@mikooomich I don't know if you remember man, I am still waiting for the sketch!

mikooomich commented:
Hey, sorry about that, had a few things come up. Here is my rough sketch. I'll admit, I'm not that good of a designer myself, but this is what I feel is a good balance between usability and being realistic to code/maintain. Of course, feel free to take artistic liberty as you wish
<img width="1920" height="1080" alt="mock playlist" src="https://github.com/user-attachments/assets/888bb0c1-7bcd-45c5-8f66-880fa13eee6f" />

TheStef56 commented:
> Hey, sorry about that, had a few things come up. Here is my rough sketch. I'll admit, I'm not that good of a designer myself, but this is what I feel is a good balance between usability and being realistic to code/maintain. Of course, feel free to take artistic liberty as you wish <img alt="mock playlist" width="1920" height="1080" src="https://private-user-images.githubusercontent.com/50182620/499927028-888bb0c1-7bcd-45c5-8f66-880fa13eee6f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAxMTc1MDksIm5iZiI6MTc2MDExNzIwOSwicGF0aCI6Ii81MDE4MjYyMC80OTk5MjcwMjgtODg4YmIwYzEtN2JjZC00NWM1LThmNjYtODgwZmExM2VlZTZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDEwVDE3MjY0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzMzg5YWViMTk3YzQ2MGE4ZmVmYTA0ZTIwZDEwMTRjMDEwMDcwYTRhYzUwZDM3YjNmM2M4MGM4NGY1YjFiMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2IsoShW4i62tzeLCHaff3QarG00oYFNp_wDjIIXgz6s">

Ok, got it. I also thought to add search bar in a playlist-style to let the user browse his imported song (to check maybe if a particular one is there or wrong) and to fill the empty space a bit. Then I would place Icon + title directly below (maybe with a little bit of description) and then the rest of the UI you suggested. What do you think?

mikooomich commented:
yep, sure

TheStef56 commented:
@mikooomich I think that is good enough, what do you think? (I changed the video preview)

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1138 @@
+                                                            Pair(prevSongQuery, item),
+                                                            UUID.randomUUID().toString(),
+                                                            true
+                                                        ) as Triple<Pair<String, Song>, String, Boolean> // TODO: m3u: why
```
Why is the type mismatched?  Pair<String, Song> vs Pair<Pair<String, Song>, Song>

TheStef56 replied:
forgot that prevSongQuery was a Pair -_-. fixed

mikooomich commented on app/src/main/java/com/dd3boh/outertune/viewmodels/ImportM3uViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+    database: MusicDatabase,
+): ViewModel() {
+    val scope = CoroutineScope(Dispatchers.IO)
+    val importedSongs = mutableStateListOf<Triple<Pair<String, Song>, String, Boolean>>()
```
Please make this a data class instead. It will be much easier to maintain in the future

Instead of calling viewmodel.importedSongs constantly on the ui side, you can change this to a flow and do .collectAsState(). You can look at LocalPlaylistScreen as an example, it's probably easier to support search and multiselect that way too.

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+    val onExitSelectionMode = {
+        inSelectMode = false
+        selection.clear()
+    }
```
Optional to implement at this moment

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+
+    val importScope = remember {
+        CoroutineScope(importJob + Dispatchers.IO)
+    }
```
Why not just use the lm scanner scope?

TheStef56 replied:
I was testing, forgot to put back. fixed.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+                                    },
+                                    enabled = !isLoading
+                                ) {
+                                    Text("import m3u") // TODO: add to R.string
```
Reminder to do string resources

TheStef56 replied:
done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
Note:

For replace search: The search query is updated in the view model, and local database search is done
automatically. Online search uses the same query but is done by manually calling the search function.

Local and online have their own search result variables and UIs.

Plz read view model comments before this file so it makes more sense

TheStef56 replied:
ok ok, thanks

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+                            .filter {
+                                when (importedChipsValue) {
+                                    importM3uFilter.IMPORTED -> it.third
+                                    importM3uFilter.REJECTED -> !it.third
```
importM3uFilter.REJECTED -> ImportM3uFilter.MISSING may be better?

Enums should start with capital.

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+                                                        Pair(prevSongQuery, item),
+                                                        UUID.randomUUID().toString(),
+                                                        true
+                                                    ) as Triple<Pair<String, Song>, String, Boolean> // TODO: m3u: why
```
Here Pair<String, Song> is actually.... Pair<Pair<String, Song>, Song> ??? As I said, data class makes much more sense here

TheStef56 replied:
fixed

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1098 @@
+    }
+
+
+    if (showEditOptions) {
```
I would rather this be a bottom sheet menu

TheStef56 replied:
<img width="405" height="894" alt="screenshot" src="https://github.com/user-attachments/assets/9ac6f0d5-934f-4b90-b979-39616d0131e0" />

I implemented it like this, with the two fields (swap and delete) added at the bottom. Tell me if there is any option that, in your opinion, has no sense to be present in this screen.

mikooomich replied:
Please make this its own menu component. You can keep swap, delete (maybe rename this so we clarify its removing from the import), details, view artist, view album, (and share ig), and remove the rest. As long as it doesn't mess up the flow of playlist import, its fine imo.

TheStef56 replied:
ok 👌🏻

mikooomich reviewed (changes requested):
I integrated the scan options beck into the main screen. It's probably makes the most sense so users won't have to change settings on every import. 

Changed to navhost to flip between the search page, to avoid having to pass replaceSong through several screens from MainActivity. Lmk your thoughts about my changes too :)

TheStef56 reviewed (commented):
Everything looks, good, i'll merge and fix/change the rest

TheStef56 reviewed (commented):
ok

TheStef56 commented:
I'm also thinking to add "wrong?" or something like that to the chips, to filter for songs that have mismatched title or artists, to let the user change them altogether.

mikooomich commented:
> I'm also thinking to add "wrong?" or something like that to the chips, to filter for songs that have mismatched title or artists, to let the user change them altogether.

Ah, yeah. Perhaps call it... "potential mismatch"? 

I forgot to mention this, but the problematic songs should be highlighted. Ex missing has a material theme error background, mismatch is some orange-yellow shade?

Btw, I squashed your commits into the first commit, so you didnt have to merge, next time it would be better to `git reset --hard origin/different_ui` when a squash happens

TheStef56 commented:
> > I'm also thinking to add "wrong?" or something like that to the chips, to filter for songs that have mismatched title or artists, to let the user change them altogether.
> 
> Ah, yeah. Perhaps call it... "potential mismatch"?
> 
> I forgot to mention this, but the problematic songs should be highlighted. Ex missing has a material theme error background, mismatch is some orange-yellow shade?
> 
> Btw, I squashed your commits into the first commit, so you didnt have to merge, next time it would be better to `git reset --hard origin/different_ui` when a squash happens

Oh okay, thanks. I'll proceed.

TheStef56 commented:
@mikooomich I did everything, except for selection mode. Can you see the result and tell me if there is something that you like or don't like?

mikooomich commented:
Thanks, I'll check it out in the coming week or so

mikooomich commented on app/src/main/res/values/strings.xml:
```diff
@@ -93,6 +101,8 @@ We recycle translations from upstream, they are pulled periodically. -->
     <string name="add_all_to_library">Add all to library</string>
     <string name="remove_from_library">Remove from library</string>
+    <string name="swap_song">Swap song</string>
+    <string name="delete_import">Remove import</string>
```
These should go in the library manager section. Maybe merge into <!-- M3uImport --> and change that to <!-- Library manager -->

TheStef56 replied:
done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/viewmodels/OnlineSearchViewModel.kt:
Why this change? ImportMenuViewModel is what we're using

TheStef56 replied:
I forgot to revert it back when whe changed the implementation. Done.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/menu/PlayerMenu.kt:
Plz remove these asserted not nulls, the null is handled above

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -174,6 +176,9 @@ fun ImportM3uScreen(
     viewModel: ImportM3uViewModel = hiltViewModel()
 ) {
+    val redError = Color(0.949f, 0.188f, 0.133f)
+    val orangeWarning = Color(0.961f, 0.682f, 0.443f)
```
Not using MaterialTheme.colorScheme.errorContainer if it looks bad is ok, but you need different shades for light and dark themes, or else the readability is questionable. Why not instead:

red = MaterialTheme.colorScheme.errorContainer.copy( boost red channel to make it brighter?)
orange = MaterialTheme.colorScheme.errorContainer.copy( brighten the green channel or something to make it orange lol)

And you'll get a naturally lighter background in light, darker in dark.

TheStef56 replied:
it's not about shades...in dark mode the red becomes really pale and bright, while the light orange becomes red. I tried MaterialTheme.colorScheme.
error
onError
containerError
onContaonererror
but they all behave the same.

mikooomich replied:
The current issue with this is readability. That's the point of those shades (well, at least how google intended it, even if it looks odd).   Light mode needs lighter background, dark mode needs darker background due to the black/white text. You can't use a single colour for both: 
<img width="300" alt="image" src="https://github.com/user-attachments/assets/a8a24b5c-18d3-40e6-a392-35ad9e6dd536" />

mikooomich commented on app/src/main/res/values/strings.xml:
```diff
@@ -58,6 +58,14 @@ We recycle translations from upstream, they are pulled periodically. -->
+    <string name="import_m3u">Import m3u</string>
+    <string name="filter_imported">Imported</string>
+    <string name="filter_missing">Missing</string>
+    <string name="filter_mismatch">Mismatch</string>
```
Ok ig, but we should leave it up the translaters

 <string name="song_added_to_queue">Imported (\"%s\")</string>

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/viewmodels/ImportM3uViewModel.kt:
```diff
@@ -79,4 +80,10 @@ class ImportM3uViewModel @Inject constructor(
+}
+
+data class ImportedSong(
+    val querySong: Pair<String, Song>,
```
If this is just one song, no need for a pair, just use two fields

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -516,7 +522,7 @@ fun ImportM3uScreen(
                                 }
                                 Text(
-                                    text = "Import a playlist to get\nstarted", // TODO: add to R.string
+                                    text = stringResource(R.string.import_playlist_to_get_started), // TODO: add to R.string
```
Did you forget to remove the TODO?

TheStef56 replied:
Yep

mikooomich reviewed (changes requested):
Hello. There are a few more things, and then we should be good after that. Thanks again!

- Crash when click view artist, view album (are you passing the rigth navcontroller)
- Importing empty playlists causes loading at 100%
- "Are you sure you want to exit" Dialog would be nice to prevent the user from accidentally closing the screen
- Toggle like in the menu does not work

TheStef56 commented:
> Hello. There are a few more things, and then we should be good after that. Thanks again!
> 
> * Crash when click view artist, view album (are you passing the rigth navcontroller)
> * Importing empty playlists causes loading at 100%
> * "Are you sure you want to exit" Dialog would be nice to prevent the user from accidentally closing the screen
> * Toggle like in the menu does not work

I fixed everything. But for me, an empty playlist with just #EXTM3U at the beginning, would show loading at 0% (not 100%) , until you would dismiss the snackbar notification (or until it would disappear by itself). I put the notification in a coroutine to solve it.

TheStef56 commented:
@mikooomich I was noticing that now you removed the ability to play the songs that are in the importM3uList or importM3uSearch routes. I think that was an important feature, since one could listen to them to enusure that the songs they are importing are actually the ones they want to. Songs can still be played by using LocalPlayerConnection.current, but the BottomPlayerSheet is not visible, since it's rendered only in the main navHost for the main navController, not in the navHost created for the m3uNavController. Should I slap a MiniPlayer and relative BottomSheet on the bottom of ImporM3uPlayerScreen (would require copy-pasting 800+ lines of code, wouldn't be a refactor much better?) or should we merge the routes in the main navHost and use just one navController? Tell me if I misunderstood something please.

Edit: I'm also working on multiple-song selection, to delete them in batches (if the user want). Maybe next I will try to implement multiple songs swap (for example you select a bunch, go to swap and you see the song you are swapping on the top, once you swap it with the one you searched, the next one is shown, you search another song, replace it, and so on)?

mikooomich commented:
> BottomPlayerSheet is not visible, since it's rendered only in the main navHost for the main navController, not in the navHost created for the m3uNavController

It should be visable, the m3u navcontroller is nested in the main one. you just need to allow it to show (in MainActivity)

> Should I slap a MiniPlayer and relative BottomSheet on the bottom of ImporM3uPlayerScreen

No. The easy thing to do is play a queue with a hardcoded stringresource name like "M3U preview" or something. That way you can reuse the existing player and queue without destoying the previous playing queue. The queue and player should be already set up to do that, lmk if you hit a snag.


>  multiple songs swap 
I don't think this is a good idea, might over-complicate things


> I fixed everything. But for me, an empty playlist with just #EXTM3U at the beginning, would show loading at 0% (not 100%)
Ah, mine was just an empty file with a .m3u extension

TheStef56 commented:
@mikooomich I enabled BottomPlayerSheet for M3uList and M3uSearch routes, letting the user play the songs in the two lists by clicking on them. I also implemented selection mode, but currently the only option is delete. Tell me if there is something that should be changed, otherwise I think this pr can finally be merged.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/component/SelectHeader.kt:
Don't do this. Select header is a universal component. I'd much rather you copy the header into a new composable for m3u

TheStef56 replied:
done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1223 @@
+                                                    } else {
+                                                        playerConnection?.playQueue(
+                                                            ListQueue(
+                                                                items = List(1) { song.toMediaMetadata() },
```
listOf(song.toMediaMetadata())

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1223 @@
+                                                            } else {
+                                                                playerConnection?.playQueue(
+                                                                    ListQueue(
+                                                                        items = List(1) { (item as SongItem).toMediaMetadata() },
```
listOf(). Also, please check the cast

TheStef56 replied:
Done

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/ImportM3uScreen.kt:
```diff
@@ -0,0 +1,1223 @@
+                onDeselectAll = { selection.clear() },
+                menuState = menuState,
+                onDismiss = onExitSelectionMode,
+                importM3uList = Pair(
```
this can be one line

TheStef56 replied:
Done

mikooomich reviewed (commented):
when in selection mode, clicking on the song plays instead of selecting it, this is inconsistent with the rest of the app (ex songs, folders, etc)

Selecting all selects all songs, instead of the all the songs in the current filter

mikooomich commented:
As I said before, reset instead of merging the branch. **_Do not merge the branch_**. Everytime I clean up your past merges, you merge basically nothing because Its already clean. I'm squashing everything the next time this hapens

TheStef56 commented:
> when in selection mode, clicking on the song plays instead of selecting it, this is inconsistent with the rest of the app (ex songs, folders, etc)
> 
> Selecting all selects all songs, instead of the all the songs in the current filter

Solved. Now add-all is cumulative. You can add all the songs within a filter, but they are kept if you change filter.

TheStef56 commented:
> As I said before, reset instead of merging the branch. **_Do not merge the branch_**. Everytime I clean up your past merges, you merge basically nothing because Its already clean. I'm squashing everything the next time this hapens

Yep, did it this time. Sorry, last time I forgot.

mikooomich commented:
Its probably fine now. Thanks!

btw, You may see me throw some extra commits, fix formatting, change up the colours, etc. Plz do not modify this pr.

> The current issue with this is readability. That's the point of those shades (well, at least how google intended it, even if it looks odd). Light mode needs lighter background, dark mode needs darker background due to the black/white text. You can't use a single colour for both: <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/50182620/505654011-a8a24b5c-18d3-40e6-a392-35ad9e6dd536.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0MjI4ODQsIm5iZiI6MTc2MTQyMjU4NCwicGF0aCI6Ii81MDE4MjYyMC81MDU2NTQwMTEtYThhMjRiNWMtMThkMy00MGU2LWEzOTItMzVhZDllNmRkNTM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI1VDIwMDMwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1ZTQ2N2M2MWFjYjViZmZmZDEwOGY2YWU2OGNhMjhmMWE0NTBjMDMxN2E3NTQ0ZjgyNjlhNzU1ZjM1YzM2ZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dHRMfb100GQr1oKEOodNUZzWo4ClwXSSfh7guZ5l5P8">

TheStef56 commented:
> Its probably fine now. Thanks!
> 
> btw, You may see me throw some extra commits, fix formatting, change up the colours, etc. Plz do not modify this pr.
> 
> > The current issue with this is readability. That's the point of those shades (well, at least how google intended it, even if it looks odd). Light mode needs lighter background, dark mode needs darker background due to the black/white text. You can't use a single colour for both: <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/50182620/505654011-a8a24b5c-18d3-40e6-a392-35ad9e6dd536.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0MjI4ODQsIm5iZiI6MTc2MTQyMjU4NCwicGF0aCI6Ii81MDE4MjYyMC81MDU2NTQwMTEtYThhMjRiNWMtMThkMy00MGU2LWEzOTItMzVhZDllNmRkNTM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI1VDIwMDMwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1ZTQ2N2M2MWFjYjViZmZmZDEwOGY2YWU2OGNhMjhmMWE0NTBjMDMxN2E3NTQ0ZjgyNjlhNzU1ZjM1YzM2ZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dHRMfb100GQr1oKEOodNUZzWo4ClwXSSfh7guZ5l5P8">

Ok, no problem!

About the colors: the default theme had readabilty problems too, but had also the problem that, on dark mode, the shades were inverted, resulting in a red "warning" and a light  horrible yellowish "error".

mikooomich committed:
material-color-utilities: Update to c50b4106030b35449baf7ec71b74963deff072dd

mikooomich committed:
ui: Fix dynamic theme off contrast

Fixes #981
Fixes #979
Fixes #978
Fixes #977

mikooomich commented:
Notes for me in the future:
- Main conflicts due to change in show nav/player changes in recomposition wars. I don't have a sane way to show/hide nav player...
- Do recomposition wars here too

mikooomich committed:
ui: The recomposition crusades have hit the player screen!

mikooomich committed:
ui: No double overlay for player gradient/blur

mikooomich committed:
ui: Resurrect tablet UI

mikooomich committed:
wip: ui: Player screen

mikooomich committed:
ui: Scrollable dialogs

mikooomich committed:
app: gradle: Optimize imports, bump dependencies

mikooomich committed:
app: Misc

mikooomich committed:
ui: Better fix for dynamic theme off contrast + high contrast mode

Fixes #981
Fixes #979
Fixes #978
Fixes #977

mikooomich committed:
outertune: Version bump to 0.10.1-b1

kubakalisciak committed:
Translated using Weblate (Polish)

Currently translated at 97.6% (253 of 259 strings)

Co-authored-by: Kuba Kaliściak <jakub.kalisciak@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/pl/
Translation: OuterTune/OuterTune-specific strings

happysurf committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: happysurf <stemen1@yahoo.it>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/it/
Translation: OuterTune/OuterTune-specific strings

yriko62 committed:
Translated using Weblate (Indonesian)

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: Muhammad Ikrawm Mubarok <hmdiandra@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/id/
Translation: OuterTune/OuterTune-specific strings

Commit:
Translated using Weblate (Estonian)

Currently translated at 54.0% (140 of 259 strings)

Co-authored-by: Priit Jõerüüt <jrthwlate@users.noreply.hosted.weblate.org>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/et/
Translation: OuterTune/OuterTune-specific strings

lucasmz-dev committed:
Added translation using Weblate (Filipino)

Co-authored-by: LucasMZ <git@lucasmz.dev>

Gwocha committed:
Translated using Weblate (Russian)

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: korablin <218695736+Gwocha@users.noreply.github.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ru/
Translation: OuterTune/OuterTune-specific strings

griperin committed:
Translated using Weblate (Catalan)

Currently translated at 20.8% (54 of 259 strings)

Co-authored-by: Griperin <griperin12@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ca/
Translation: OuterTune/OuterTune-specific strings

tranquocvinh2828-glitch committed:
Translated using Weblate (Vietnamese)

Currently translated at 46.3% (120 of 259 strings)

Co-authored-by: Tran Quoc Vinh <tranquocvinh2828@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/vi/
Translation: OuterTune/OuterTune-specific strings

papapep committed:
Translated using Weblate (Catalan)

Currently translated at 22.3% (58 of 259 strings)

Co-authored-by: papapep <papapep@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ca/
Translation: OuterTune/OuterTune-specific strings

mikooomich committed:
player: Migrate all direct service accesses to playerconnection

Close #895

* Hopefully fixes player/queue issue when app is freshly launched

mikooomich committed:
lm: Remove local to remote artist linking

mikooomich committed:
lm: Scanners: Improve song linking consistency

lucasmz-dev committed:
translation: update upstream strings (#987)

as per commit bfba5ecb16795cfc93ecce73d94ca583bcec917c

Co-authored-by: BananaChad <owencsousa@gmail.com>
Co-authored-by: Dwip Biswas <dwip@dwip.dedyn.io>

mikooomich committed:
app: Remove update checker

mikooomich committed:
outetune: Update readme

xusism committed:
Translated using Weblate (Catalan)

Currently translated at 22.3% (58 of 259 strings)

Translated using Weblate (Spanish)

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: Xusi Fons <xusi@keemail.me>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ca/
Translate-URL: https://hosted.weblate.org/projects/outertune/app/es/
Translation: OuterTune/OuterTune-specific strings

davex536 committed:
Translated using Weblate (Polish)

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: Dawid <davidosbananos5@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/pl/
Translation: OuterTune/OuterTune-specific strings

pxsit committed:
Translated using Weblate (Thai)

Currently translated at 17.3% (45 of 259 strings)

Co-authored-by: Pasit Sangprachathanarak <ouipingpasit@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/th/
Translation: OuterTune/OuterTune-specific strings

smarquespt committed:
Translated using Weblate (Portuguese)

Currently translated at 99.2% (257 of 259 strings)

Translated using Weblate (Portuguese)

Currently translated at 99.2% (257 of 259 strings)

Translated using Weblate (Portuguese)

Currently translated at 99.2% (257 of 259 strings)

Translated using Weblate (Portuguese)

Currently translated at 99.2% (257 of 259 strings)

Co-authored-by: Sérgio Marques <smarquespt@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/pt/
Translation: OuterTune/OuterTune-specific strings

Commit:
Translated using Weblate (Japanese)

Currently translated at 96.1% (249 of 259 strings)

Co-authored-by: S-H-Y-A <s-h-y-a@users.noreply.hosted.weblate.org>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ja/
Translation: OuterTune/OuterTune-specific strings

Atalanttore committed:
Translated using Weblate (German)

Currently translated at 95.7% (248 of 259 strings)

Co-authored-by: Ettore Atalan <atalanttore@googlemail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/de/
Translation: OuterTune/OuterTune-specific strings

weblate committed:
Update translation files

Updated by "Remove blank strings" add-on in Weblate.

Co-authored-by: Hosted Weblate <hosted@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/
Translation: OuterTune/OuterTune-specific strings

Aeoniixx committed:
Translated using Weblate (Hindi)

Currently translated at 89.1% (231 of 259 strings)

Translated using Weblate (Hindi)

Currently translated at 79.5% (206 of 259 strings)

Co-authored-by: Aeoniixx <aeoniixx@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/hi/
Translation: OuterTune/OuterTune-specific strings

Commit:
Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 100.0% (259 of 259 strings)

Co-authored-by: Sketch6580 <sketch6580@users.noreply.hosted.weblate.org>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/zh_Hans/
Translation: OuterTune/OuterTune-specific strings

Drbsy committed:
Translated using Weblate (Arabic)

Currently translated at 80.6% (209 of 259 strings)

Co-authored-by: Drbsy <ouffata.ishak@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/ar/
Translation: OuterTune/OuterTune-specific strings

Commit:
Translated using Weblate (Estonian)

Currently translated at 62.1% (161 of 259 strings)

Co-authored-by: Priit Jõerüüt <jrthwlate@users.noreply.hosted.weblate.org>
Translate-URL: https://hosted.weblate.org/projects/outertune/app/et/
Translation: OuterTune/OuterTune-specific strings

mikooomich committed:
app: Bump deps, clean up imports

mikooomich committed:
player: Better way to disabled gapless offload

mikooomich committed:
app: Fix compiling release

mikooomich committed:
outertune: Version bump to 0.10.2-b1

* 71 is the version code for latest 0.10.1 release... which is a hotfix over 0.10.0, and does not contain many of the changes in 0.10.1-b1. Let's start fresh with 0.10.2 ig

lucasmz-dev committed:
translation: update upstream strings (#1088)

as per commit 4db524a0cb4a283a9434698a42bf0a76967c68c5

Co-authored-by: AriyA66Y <ariya66y@protonmail.com>
Co-authored-by: Azazel <cvakshay764@gmail.com>
Co-authored-by: bitmetax <ajaykaranmp@gmail.com>
Co-authored-by: DA D <daramgakaram@gmail.com>
Co-authored-by: Daniel Cazorla <1790251@uab.cat>
Co-authored-by: grafspirt <grafspirt@gmail.com>
Co-authored-by: Maite Guix <maite.guix@gmail.com>
Co-authored-by: Michael Moroni <michaelmoroni@disroot.org>
Co-authored-by: Mickaël Binos <mickaelbinos@outlook.com>
Co-authored-by: Stefanos Stefanidis <www.fe32gr23@gmail.com>
Co-authored-by: Yatrik Patel <fallers_quieter0v@icloud.com>

mikooomich committed:
Update README

TheStef56 commented:
I was working on my forks and deleted different_ui branch by mistake, lol. I restored everyting.
